llvm/lib/SYCLLowerIR/SYCLJointMatrixTransform.cpp

```diff
@@ -184,12 +188,73 @@ bool transformAccessChain(Function *F) {
   }
   return ModuleChanged;
 }
+
+// Per SPIR-V specification Layout of a matrix must be a constant instruction
+// aka a constexpr or specialization constant. Meanwhile in SYCL headers
+// layout is passed as a parameter to joint_matrix_load function, so even if
+// that layout is a constant expression in the user's code - it's not possible
+// to prove that to the compiler, so constant propagation will happen only
+// after inlining, not in AST. That means, that with O0 layout would remain
+// to be a runtime variable in LLVM IR.
+// SYCL matrix layout is being mapped on SPIR-V matrix layout by
+// joint_matrix_layout_to_spv function. The following routine finds calls to
+// this function and replaces them with the found constant.
+// This function also cleans up code, that becomes dead. Pattern of the dead
+// code is stable, as user's code doesn't affect it.
+bool propagateConstexprLayout(Function *F) {
```

Comment:
please also note, that I'm not even trying to debug information here. And it should be fine, as user's don't need to debug joint_matrix_layout_to_spv and other internal functions